File: 백준/Platinum/2887. 행성 터널/2887.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int parent[100001];
vector<pair<int, int>> vx;
vector<pair<int, int>> vy;
vector<pair<int, int>> vz;

int find_parent(int parent[], int x)
{
	if (parent[x] != x)
		parent[x] = find_parent(parent, parent[x]);
	return parent[x];
}

void union_parent(int parent[], int a, int b)
{
	a = find_parent(parent, a);
	b = find_parent(parent, b);

	if (a < b)
		parent[b] = a;
	else
		parent[a] = b;
}

int main()
{
	int n;
	cin >> n;

	for (int i = 0; i < n; i++)
		parent[i] = i;

	for (int i = 0; i < n; i++)
	{
		int x, y, z;
		cin >> x >> y >> z;

		vx.push_back(make_pair(x, i));
		vy.push_back(make_pair(y, i));
		vz.push_back(make_pair(z, i));
	}

	sort(vx.begin(), vx.end());
	sort(vy.begin(), vy.end());
	sort(vz.begin(), vz.end());

	priority_queue<pair<int, pair<int, int>>> pq;
	for (int i = 0; i < n - 1; i++)
	{
		pq.push({ -abs(vx[i].first - vx[i + 1].first), {vx[i].second, vx[i + 1].second } });
		pq.push({ -abs(vy[i].first - vy[i + 1].first), {vy[i].second, vy[i + 1].second } });
		pq.push({ -abs(vz[i].first - vz[i + 1].first), {vz[i].second, vz[i + 1].second } });
	}

	int result = 0;
	while (!pq.empty()) 
	{
		int cost = -pq.top().first;
		int a = pq.top().second.first;
		int b = pq.top().second.second;
		pq.pop();

		if (find_parent(parent, a) != find_parent(parent, b))
		{
			union_parent(parent, a, b);
			result += cost;

			//cout << a << " " << b << " " << cost << '\n';
		}
	}
	cout << result;
}

File: 백준/Platinum/2887. 행성 터널/2887.java
import java.util.*;
import java.io.*;

public class Main {

    static int[] parent;
    static List<int[]> vx = new ArrayList<>();
    static List<int[]> vy = new ArrayList<>();
    static List<int[]> vz = new ArrayList<>();

    static int findParent(int x) {
        if (parent[x] != x)
            parent[x] = findParent(parent[x]);
        return parent[x];
    }

    static void unionParent(int a, int b) {
        a = findParent(a);
        b = findParent(b);
        if (a < b) parent[b] = a;
        else parent[a] = b;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        parent = new int[n];
        for (int i = 0; i < n; i++)
            parent[i] = i;

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            int z = Integer.parseInt(st.nextToken());

            vx.add(new int[]{x, i});
            vy.add(new int[]{y, i});
            vz.add(new int[]{z, i});
        }

        vx.sort(Comparator.comparingInt(a -> a[0]));
        vy.sort(Comparator.comparingInt(a -> a[0]));
        vz.sort(Comparator.comparingInt(a -> a[0]));

        // priority queue (min-heap) - smallest cost first
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));

        for (int i = 0; i < n - 1; i++) {
            pq.offer(new int[]{Math.abs(vx.get(i)[0] - vx.get(i + 1)[0]), vx.get(i)[1], vx.get(i + 1)[1]});
            pq.offer(new int[]{Math.abs(vy.get(i)[0] - vy.get(i + 1)[0]), vy.get(i)[1], vy.get(i + 1)[1]});
            pq.offer(new int[]{Math.abs(vz.get(i)[0] - vz.get(i + 1)[0]), vz.get(i)[1], vz.get(i + 1)[1]});
        }

        int result = 0;
        while (!pq.isEmpty()) {
            int[] edge = pq.poll();
            int cost = edge[0];
            int a = edge[1];
            int b = edge[2];

            if (findParent(a) != findParent(b)) {
                unionParent(a, b);
                result += cost;
                // System.out.println(a + " " + b + " " + cost);
            }
        }

        System.out.println(result);
    }
}
